aip/general/0161.md

```diff
@@ -0,0 +1,131 @@
+---
+id: 161
+state: reviewing
+created: 2021-03-01
+placement:
+  category: design-patterns
+  order: 85
+---
+
+# Field masks
+
+Often, when updating resources (using an update method as defined in AIP-134 or
+something reasonably similar), it is desirable to specify exactly which fields
+are being updated, so that the service can ignore the rest, even if the user
+sends new values.
+
+It is tempting to define a mask format to handle the precise needs for each
+API. However, masking requirements evolve, and therefore it is prudent to use a
+structured syntax. This allows updates to be able to be made transparently,
+without waiting for UI or client updates.
+
+## Guidance
+
+Update methods (AIP-134) **must** include an `update_mask` field, and Get and
+List methods **may** include a `read_mask` field. These are collectively called
+"field masks", and they use the `google.protobuf.FieldMask` type.
+
+Field masks **must** always be relative to the resource:
+
+```proto
+message UpdateBookRequest {
+  // The book to update.
+  //
+  // The book's `name` field is used to identify the book to be updated.
+  // Format: publishers/{publisher}/books/{book}
+  Book book = 1 [(google.api.field_behavior) = REQUIRED];
+
+  // The list of fields to be updated.
+  // Fields are specified relative to the book
+  // (e.g. `title`, `rating`; *not* `book.title` or `book.rating`).
+  google.protobuf.FieldMask update_mask = 2;
+}
+```
+
+### Read-write consistency
+
+Read and write behavior for field masks **must** be self-consistent if a mask
+is present:
+
+- If a user updates a resource with a given mask, and then reads the same
+  resource with the same mask, the exact same data **must** be returned.
+- Similarly, reading a resource with a given mask and then updating the
+  resource with the returned data and the same mask **must** be a no-op.
+
+**Note:** This implies that any mask that is valid for either read or write
```

Comment:
That seems odd - it suggests you can't include output-only fields in a field mask.
If I want to get (say) "name" and "create_time" (and only those fields), how do I do that?

Reply:
Why does it imply this? (Legitimately confused.)

Reply:
Well, let's assume that `name,create_time` is a valid field mask for read... it implies that it should be valid for write as well. But does it make sense to specify a field mask for a field that you can't write? I'd assumed that would be invalid. (It certainly gives the impression that the user is *trying* to intentionally write that field - and it's normally nice to tell users when they're trying to do something they're not actually going to be able to do.)

Reply:
I would certainly think that an update mask that _included_ an output only field (e.g. `*` or a message that has an output-only sub-field) should be permitted.

In the past, we have told service owners to permit update masks that explicitly included output only fields also, although I think that point is shakier, although I see this question as an argument in its favor, because I want this code to be permitted:

```typescript
let mask = ['name', 'rating', 'create_time']
for (let name: string of [ ... ]) {
  let book = client.getBook(name, {'read_mask': mask})
  // All classics get a bump in rating because they stood the test of time.
  if (book.create_time < new Date(1900, 1, 1)) {
    book.rating += 1
    client.updateBook(book, {'update_mask': mask})
  }
}
```

Reply:
Right. That all makes sense to me. I think it would be good to note that explicitly. (It can be in a later PR though.)